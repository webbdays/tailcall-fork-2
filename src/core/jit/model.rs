use std::collections::HashMap;
use std::fmt::{Debug, Formatter};

use async_graphql::parser::types::{ConstDirective, OperationType};
use async_graphql::{Name, Pos, Positioned};
use async_graphql_value::ConstValue;
use serde::Deserialize;

use crate::core::ir::model::IR;

#[derive(Debug, Deserialize, Clone)]
pub struct Variables<Value>(HashMap<String, Value>);

impl<Value> Default for Variables<Value> {
    fn default() -> Self {
        Self::new()
    }
}

impl<Value> Variables<Value> {
    pub fn new() -> Self {
        Self(HashMap::new())
    }
    pub fn get(&self, key: &str) -> Option<&Value> {
        self.0.get(key)
    }
    pub fn insert(&mut self, key: String, value: Value) {
        self.0.insert(key, value);
    }
    pub fn try_map<Output, Error>(
        self,
        map: impl Fn(Value) -> Result<Output, Error>,
    ) -> Result<Variables<Output>, Error> {
        let mut hm = HashMap::new();
        for (k, v) in self.0 {
            hm.insert(k, map(v)?);
        }
        Ok(Variables(hm))
    }
}

impl<V> FromIterator<(String, V)> for Variables<V> {
    fn from_iter<T: IntoIterator<Item = (String, V)>>(iter: T) -> Self {
        Self(iter.into_iter().collect())
    }
}

#[derive(Debug, Clone)]
pub struct Arg<Input> {
    pub id: ArgId,
    pub name: String,
    pub type_of: crate::core::blueprint::Type,
    pub value: Option<Input>,
    pub default_value: Option<Input>,
}

impl<Input> Arg<Input> {
    pub fn try_map<Output, Error>(
        self,
        map: impl Fn(Input) -> Result<Output, Error>,
    ) -> Result<Arg<Output>, Error> {
        Ok(Arg {
            id: self.id,
            name: self.name,
            type_of: self.type_of,
            value: self.value.map(&map).transpose()?,
            default_value: self.default_value.map(&map).transpose()?,
        })
    }
}

#[derive(Clone)]
pub struct ArgId(usize);

impl Debug for ArgId {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl ArgId {
    pub fn new(id: usize) -> Self {
        ArgId(id)
    }
}

#[derive(Clone, PartialEq, Eq, Hash)]
pub struct FieldId(usize);

impl Debug for FieldId {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl FieldId {
    pub fn new(id: usize) -> Self {
        FieldId(id)
    }
    pub fn as_usize(&self) -> usize {
        self.0
    }
}

#[derive(Clone)]
pub struct Field<Extensions, Input> {
    pub id: FieldId,
    pub name: String,
    pub ir: Option<IR>,
    pub type_of: crate::core::blueprint::Type,
    pub skip: Option<Variable>,
    pub include: Option<Variable>,
    pub args: Vec<Arg<Input>>,
    pub extensions: Option<Extensions>,
    pub pos: Pos,
    pub is_scalar: bool,
    pub directives: Vec<Directive<Input>>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct Variable(String);

impl Variable {
    pub fn new(name: String) -> Self {
        Variable(name)
    }
    pub fn as_str(&self) -> &str {
        &self.0
    }
    pub fn into_string(self) -> String {
        self.0
    }
}

impl<Input> Field<Nested<Input>, Input> {
    pub fn try_map<Output, Error>(
        self,
        map: &impl Fn(Input) -> Result<Output, Error>,
    ) -> Result<Field<Nested<Output>, Output>, Error> {
        let mut extensions = None;

        if let Some(nested) = self.extensions {
            let mut exts = vec![];
            for v in nested.0 {
                exts.push(v.try_map(map)?);
            }
            extensions = Some(Nested(exts));
        }

        Ok(Field {
            id: self.id,
            name: self.name,
            ir: self.ir,
            type_of: self.type_of,
            extensions,
            pos: self.pos,
            skip: self.skip,
            include: self.include,
            args: self
                .args
                .into_iter()
                .map(|arg| arg.try_map(map))
                .collect::<Result<_, _>>()?,
            is_scalar: false,
            directives: self
                .directives
                .into_iter()
                .map(|directive| directive.try_map(map))
                .collect::<Result<_, _>>()?,
        })
    }
}

impl<Input> Field<Flat, Input> {
    pub fn try_map<Output, Error>(
        self,
        map: impl Fn(Input) -> Result<Output, Error>,
    ) -> Result<Field<Flat, Output>, Error> {
        Ok(Field {
            id: self.id,
            name: self.name,
            ir: self.ir,
            type_of: self.type_of,
            extensions: self.extensions,
            skip: self.skip,
            include: self.include,
            pos: self.pos,
            args: self
                .args
                .into_iter()
                .map(|arg| arg.try_map(&map))
                .collect::<Result<_, _>>()?,
            is_scalar: self.is_scalar,
            directives: self
                .directives
                .into_iter()
                .map(|directive| directive.try_map(&map))
                .collect::<Result<_, _>>()?,
        })
    }
}

impl<Input> Field<Nested<Input>, Input> {
    pub fn nested(&self) -> Option<&Vec<Field<Nested<Input>, Input>>> {
        self.extensions.as_ref().map(|Nested(nested)| nested)
    }

    pub fn nested_iter(&self) -> impl Iterator<Item = &Field<Nested<Input>, Input>> {
        self.nested()
            .map(|nested| nested.iter())
            .into_iter()
            .flatten()
    }
}

impl<Input> Field<Flat, Input> {
    fn parent(&self) -> Option<&FieldId> {
        self.extensions.as_ref().map(|Flat(id)| id)
    }

    fn into_nested(self, fields: &[Field<Flat, Input>]) -> Field<Nested<Input>, Input>
    where
        Input: Clone,
    {
        let mut children = Vec::new();
        for field in fields.iter() {
            if let Some(id) = field.parent() {
                if *id == self.id {
                    children.push(field.to_owned().into_nested(fields));
                }
            }
        }

        let extensions = if children.is_empty() {
            None
        } else {
            Some(Nested(children))
        };

        Field {
            id: self.id,
            name: self.name,
            ir: self.ir,
            type_of: self.type_of,
            skip: self.skip,
            include: self.include,
            args: self.args,
            pos: self.pos,
            extensions,
            is_scalar: self.is_scalar,
            directives: self.directives,
        }
    }
}

impl<Extensions: Debug, Input: Debug> Debug for Field<Extensions, Input> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let mut debug_struct = f.debug_struct("Field");
        debug_struct.field("id", &self.id);
        debug_struct.field("name", &self.name);
        if self.ir.is_some() {
            debug_struct.field("ir", &"Some(..)");
        }
        debug_struct.field("type_of", &self.type_of);
        if !self.args.is_empty() {
            debug_struct.field("args", &self.args);
        }
        if self.extensions.is_some() {
            debug_struct.field("extensions", &self.extensions);
        }
        debug_struct.field("is_scalar", &self.is_scalar);
        if self.skip.is_some() {
            debug_struct.field("skip", &self.skip);
        }
        if self.include.is_some() {
            debug_struct.field("include", &self.include);
        }
        debug_struct.field("directives", &self.directives);
        debug_struct.finish()
    }
}

/// Stores field relationships in a flat structure where each field links to its
/// parent.
#[derive(Clone)]
pub struct Flat(FieldId);

impl Flat {
    pub fn new(id: FieldId) -> Self {
        Flat(id)
    }
    pub fn as_id(&self) -> &FieldId {
        &self.0
    }
}
impl Debug for Flat {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "Flat({:?})", self.0)
    }
}

/// Store field relationships in a nested structure like a tree where each field
/// links to its children.
#[derive(Clone, Debug)]
pub struct Nested<Input>(Vec<Field<Nested<Input>, Input>>);

#[derive(Clone, Debug)]
pub struct OperationPlan<Input> {
    flat: Vec<Field<Flat, Input>>,
    operation_type: OperationType,
    nested: Vec<Field<Nested<Input>, Input>>,
}

impl<Input> OperationPlan<Input> {
    pub fn try_map<Output, Error>(
        self,
        map: impl Fn(Input) -> Result<Output, Error>,
    ) -> Result<OperationPlan<Output>, Error> {
        let mut flat = vec![];

        for f in self.flat {
            flat.push(f.try_map(&map)?);
        }

        let mut nested = vec![];

        for n in self.nested {
            nested.push(n.try_map(&map)?);
        }

        Ok(OperationPlan { flat, operation_type: self.operation_type, nested })
    }
}

impl<Input> OperationPlan<Input> {
    pub fn new(fields: Vec<Field<Flat, Input>>, operation_type: OperationType) -> Self
    where
        Input: Clone,
    {
        let nested = fields
            .clone()
            .into_iter()
            .filter(|f| f.extensions.is_none())
            .map(|f| f.into_nested(&fields))
            .collect::<Vec<_>>();

        Self { flat: fields, nested, operation_type }
    }

    pub fn operation_type(&self) -> OperationType {
        self.operation_type
    }

    pub fn is_query(&self) -> bool {
        self.operation_type == OperationType::Query
    }

    pub fn as_nested(&self) -> &[Field<Nested<Input>, Input>] {
        &self.nested
    }

    pub fn into_nested(self) -> Vec<Field<Nested<Input>, Input>> {
        self.nested
    }

    pub fn as_parent(&self) -> &[Field<Flat, Input>] {
        &self.flat
    }

    pub fn find_field(&self, id: FieldId) -> Option<&Field<Flat, Input>> {
        self.flat.iter().find(|field| field.id == id)
    }

    pub fn find_field_path<S: AsRef<str>>(&self, path: &[S]) -> Option<&Field<Flat, Input>> {
        match path.split_first() {
            None => None,
            Some((name, path)) => {
                let field = self.flat.iter().find(|field| field.name == name.as_ref())?;
                if path.is_empty() {
                    Some(field)
                } else {
                    self.find_field_path(path)
                }
            }
        }
    }

    pub fn size(&self) -> usize {
        self.flat.len()
    }
}

#[derive(Clone, Debug)]
pub struct Directive<Input> {
    pub name: String,
    pub arguments: Vec<(String, Input)>,
}

impl<Input> Directive<Input> {
    pub fn try_map<Output, Error>(
        self,
        map: impl Fn(Input) -> Result<Output, Error>,
    ) -> Result<Directive<Output>, Error> {
        Ok(Directive {
            name: self.name,
            arguments: self
                .arguments
                .into_iter()
                .map(|(k, v)| map(v).map(|mapped_value| (k, mapped_value)))
                .collect::<Result<Vec<_>, _>>()?,
        })
    }
}

impl<'a> From<&'a Directive<ConstValue>> for ConstDirective {
    fn from(value: &'a Directive<ConstValue>) -> Self {
        // we don't use pos required in Positioned struct, hence using defaults.
        ConstDirective {
            name: Positioned::new(Name::new(&value.name), Default::default()),
            arguments: value
                .arguments
                .iter()
                .map(|a| {
                    (
                        Positioned::new(Name::new(a.0.clone()), Default::default()),
                        Positioned::new(a.1.clone(), Default::default()),
                    )
                })
                .collect::<Vec<_>>(),
        }
    }
}

#[cfg(test)]
mod test {
    use async_graphql::parser::types::ConstDirective;
    use async_graphql_value::ConstValue;

    use super::Directive;

    #[test]
    fn test_from_custom_directive() {
        let custom_directive = Directive {
            name: "options".to_string(),
            arguments: vec![("paging".to_string(), ConstValue::Boolean(true))],
        };

        let async_directive: ConstDirective = (&custom_directive).into();
        insta::assert_debug_snapshot!(async_directive);
    }
}
