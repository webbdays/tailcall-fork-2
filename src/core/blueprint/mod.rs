mod auth;
mod blueprint;
mod compress;
mod cors;
mod definitions;
mod dynamic_value;
mod from_config;
mod index;
mod into_schema;
mod links;
mod mustache;
mod operators;
mod schema;
mod server;
pub mod telemetry;
mod timeout;
mod union_resolver;
mod upstream;

pub use auth::*;
pub use blueprint::*;
pub use cors::*;
pub use definitions::*;
pub use dynamic_value::*;
pub use from_config::*;
pub use index::*;
pub use links::*;
pub use operators::*;
pub use schema::*;
pub use server::*;
pub use timeout::GlobalTimeout;
pub use upstream::*;

use crate::core::config::{Arg, ConfigModule, Field};
use crate::core::try_fold::TryFold;

pub type TryFoldConfig<'a, A> = TryFold<'a, ConfigModule, A, String>;

pub(crate) trait TypeLike {
    fn name(&self) -> &str;
    fn list(&self) -> bool;
    fn non_null(&self) -> bool;
    fn list_type_required(&self) -> bool;
}

impl TypeLike for Field {
    fn name(&self) -> &str {
        &self.type_of
    }

    fn list(&self) -> bool {
        self.list
    }

    fn non_null(&self) -> bool {
        self.required
    }

    fn list_type_required(&self) -> bool {
        self.list_type_required
    }
}

impl TypeLike for Arg {
    fn name(&self) -> &str {
        &self.type_of
    }

    fn list(&self) -> bool {
        self.list
    }

    fn non_null(&self) -> bool {
        self.required
    }

    fn list_type_required(&self) -> bool {
        false
    }
}

pub(crate) fn to_type<T>(field: &T, override_non_null: Option<bool>) -> Type
where
    T: TypeLike,
{
    let name = field.name();
    let list = field.list();
    let list_type_required = field.list_type_required();
    let non_null = if let Some(non_null) = override_non_null {
        non_null
    } else {
        field.non_null()
    };

    if list {
        Type::ListType {
            of_type: Box::new(Type::NamedType {
                name: name.to_string(),
                non_null: list_type_required,
            }),
            non_null,
        }
    } else {
        Type::NamedType { name: name.to_string(), non_null }
    }
}
