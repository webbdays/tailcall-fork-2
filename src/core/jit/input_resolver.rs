use async_graphql_value::{ConstValue, Value};

use super::{OperationPlan, ResolveInputError, Variables};

/// Trait to represent conversion from some dynamic type (with variables)
/// to the resolved variant based on the additional provided info.
/// E.g. conversion from [async_graphql_value::Value] ->
/// [async_graphql_value::ConstValue]
pub trait InputResolvable {
    type Output;

    fn resolve(
        self,
        variables: &Variables<Self::Output>,
    ) -> Result<Self::Output, ResolveInputError>;
}

impl InputResolvable for Value {
    type Output = ConstValue;

    // TODO:
    // - provide default values
    fn resolve(self, variables: &Variables<ConstValue>) -> Result<Self::Output, ResolveInputError> {
        self.into_const_with(|name| {
            variables
                .get(&name)
                .cloned()
                .ok_or_else(|| ResolveInputError::VariableIsNotFound(name.to_string()))
        })
    }
}

/// Transforms [OperationPlan] values the way that all the input values
/// are transformed to const variant with the help of [InputResolvable] trait
pub struct InputResolver<Input> {
    plan: OperationPlan<Input>,
}

impl<Input> InputResolver<Input> {
    pub fn new(plan: OperationPlan<Input>) -> Self {
        Self { plan }
    }
}

impl<Input, Output> InputResolver<Input>
where
    Input: Clone,
    Output: Clone,
    Input: InputResolvable<Output = Output>,
{
    pub fn resolve_input(
        &self,
        variables: &Variables<Output>,
    ) -> Result<OperationPlan<Output>, ResolveInputError> {
        let new_fields = self
            .plan
            .as_parent()
            .iter()
            .map(|field| field.clone().try_map(|value| value.resolve(variables)))
            .collect::<Result<_, _>>()?;

        Ok(OperationPlan::new(new_fields, self.plan.operation_type()))
    }
}
