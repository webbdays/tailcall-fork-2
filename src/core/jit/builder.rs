use std::collections::HashMap;
use std::ops::Deref;

use async_graphql::parser::types::{
    Directive, DocumentOperations, ExecutableDocument, FragmentDefinition, OperationDefinition,
    OperationType, Selection, SelectionSet,
};
use async_graphql::Positioned;
use async_graphql_value::{ConstValue, Value};

use super::input_resolver::InputResolver;
use super::model::{Directive as JitDirective, *};
use super::BuildError;
use crate::core::blueprint::{Blueprint, Index, QueryField};
use crate::core::counter::{Count, Counter};
use crate::core::jit::model::OperationPlan;
use crate::core::merge_right::MergeRight;

#[derive(PartialEq, strum_macros::Display)]
enum Condition {
    True,
    False,
    Variable(Variable),
}

struct Conditions {
    skip: Condition,
    include: Condition,
}

impl Conditions {
    /// Checks if the field should be skipped always
    fn is_const_skip(&self) -> bool {
        // Truth Table
        // skip | include | ignore
        // T   |    T    |   T
        // T   |    F    |   T
        // F   |    T    |   F
        // F   |    F    |   T
        //
        // Logical expression:
        //     say skip = p, include = q
        // (p V ~q)

        // so instead of a normalizing variables,
        // we can just check for the above condition

        matches!(
            (&self.skip, &self.include),
            (Condition::True, _) | (Condition::False, Condition::False)
        )
    }

    fn into_variable_tuple(self) -> (Option<Variable>, Option<Variable>) {
        let comp = |condition| match condition {
            Condition::Variable(var) => Some(var),
            _ => None,
        };

        let include = comp(self.include);
        let skip = comp(self.skip);

        (include, skip)
    }
}

pub struct Builder {
    pub index: Index,
    pub arg_id: Counter<usize>,
    pub field_id: Counter<usize>,
    pub document: ExecutableDocument,
}

// TODO: make generic over Value (Input) type
impl Builder {
    pub fn new(blueprint: &Blueprint, document: ExecutableDocument) -> Self {
        let index = blueprint.index();
        Self {
            document,
            index,
            arg_id: Counter::default(),
            field_id: Counter::default(),
        }
    }

    #[inline(always)]
    fn include(
        &self,
        directives: &[Positioned<async_graphql::parser::types::Directive>],
    ) -> Conditions {
        fn get_condition(dir: &Directive) -> Option<Condition> {
            let arg = dir.get_argument("if").map(|pos| &pos.node);
            match arg {
                None => None,
                Some(value) => match value {
                    Value::Boolean(bool) => {
                        let condition = if *bool {
                            Condition::True
                        } else {
                            Condition::False
                        };
                        Some(condition)
                    }
                    Value::Variable(var) => {
                        Some(Condition::Variable(Variable::new(var.deref().to_owned())))
                    }
                    _ => None,
                },
            }
        }
        Conditions {
            skip: directives
                .iter()
                .find(|d| d.node.name.node.as_str() == "skip")
                .map(|d| &d.node)
                .and_then(get_condition)
                .unwrap_or(Condition::False),
            include: directives
                .iter()
                .find(|d| d.node.name.node.as_str() == "include")
                .map(|d| &d.node)
                .and_then(get_condition)
                .unwrap_or(Condition::True),
        }
    }

    #[allow(clippy::too_many_arguments)]
    #[inline(always)]
    fn iter(
        &self,
        selection: &SelectionSet,
        type_of: &str,
        exts: Option<Flat>,
        fragments: &HashMap<&str, &FragmentDefinition>,
    ) -> Vec<Field<Flat, Value>> {
        let mut fields = vec![];
        for selection in &selection.items {
            match &selection.node {
                Selection::Field(Positioned { node: gql_field, .. }) => {
                    let conditions = self.include(&gql_field.directives);

                    // if include is always false xor skip is always true,
                    // then we can skip the field from the plan
                    if conditions.is_const_skip() {
                        continue;
                    }

                    let mut directives = Vec::with_capacity(gql_field.directives.len());
                    for directive in &gql_field.directives {
                        let directive = &directive.node;
                        if directive.name.node == "skip" || directive.name.node == "include" {
                            continue;
                        }
                        let arguments = directive
                            .arguments
                            .iter()
                            .map(|(k, v)| (k.node.to_string(), v.node.clone()))
                            .collect::<Vec<_>>();

                        directives
                            .push(JitDirective { name: directive.name.to_string(), arguments });
                    }

                    let (include, skip) = conditions.into_variable_tuple();

                    let field_name = gql_field.name.node.as_str();
                    let request_args = gql_field
                        .arguments
                        .iter()
                        .map(|(k, v)| (k.node.as_str().to_string(), v.node.to_owned()))
                        .collect::<HashMap<_, _>>();

                    if let Some(field_def) = self.index.get_field(type_of, field_name) {
                        let mut args = Vec::with_capacity(request_args.len());
                        if let QueryField::Field((_, schema_args)) = field_def {
                            for (arg_name, arg_value) in schema_args {
                                let type_of = arg_value.of_type.clone();
                                let id = ArgId::new(self.arg_id.next());
                                let name = arg_name.clone();
                                let default_value = arg_value
                                    .default_value
                                    .as_ref()
                                    .and_then(|v| v.to_owned().try_into().ok());
                                args.push(Arg {
                                    id,
                                    name,
                                    type_of,
                                    // TODO: handle errors for non existing request_args without the
                                    // default
                                    value: request_args.get(arg_name).cloned(),
                                    default_value,
                                });
                            }
                        }

                        let type_of = match field_def {
                            QueryField::Field((field_def, _)) => field_def.of_type.clone(),
                            QueryField::InputField(field_def) => field_def.of_type.clone(),
                        };

                        let id = FieldId::new(self.field_id.next());
                        let child_fields = self.iter(
                            &gql_field.selection_set.node,
                            type_of.name(),
                            Some(Flat::new(id.clone())),
                            fragments,
                        );
                        let name = gql_field
                            .alias
                            .as_ref()
                            .map(|alias| alias.node.to_string())
                            .unwrap_or(field_name.to_string());
                        let ir = match field_def {
                            QueryField::Field((field_def, _)) => field_def.resolver.clone(),
                            _ => None,
                        };
                        let flat_field = Field {
                            id,
                            name,
                            ir,
                            is_scalar: self.index.type_is_scalar(type_of.name()),
                            type_of,
                            skip,
                            include,
                            args,
                            pos: selection.pos,
                            extensions: exts.clone(),
                            directives,
                        };

                        fields.push(flat_field);
                        fields = fields.merge_right(child_fields);
                    } else {
                        // TODO: error if the field is not found in the schema
                    }
                }
                Selection::FragmentSpread(Positioned { node: fragment_spread, .. }) => {
                    if let Some(fragment) =
                        fragments.get(fragment_spread.fragment_name.node.as_str())
                    {
                        fields.extend(self.iter(
                            &fragment.selection_set.node,
                            fragment.type_condition.node.on.node.as_str(),
                            exts.clone(),
                            fragments,
                        ));
                    }
                }
                _ => {}
            }
        }

        fields
    }

    #[inline(always)]
    fn get_type(&self, ty: OperationType) -> Option<&str> {
        match ty {
            OperationType::Query => Some(self.index.get_query()),
            OperationType::Mutation => self.index.get_mutation(),
            OperationType::Subscription => None,
        }
    }

    /// Resolves currently processed operation
    /// based on [spec](https://spec.graphql.org/October2021/#sec-Executing-Requests)
    #[inline(always)]
    fn get_operation(
        &self,
        operation_name: Option<&str>,
    ) -> Result<&OperationDefinition, BuildError> {
        if let Some(operation_name) = operation_name {
            match &self.document.operations {
                DocumentOperations::Single(_) => None,
                DocumentOperations::Multiple(operations) => {
                    operations.get(operation_name).map(|op| &op.node)
                }
            }
            .ok_or_else(|| BuildError::OperationNotFound(operation_name.to_string()))
        } else {
            match &self.document.operations {
                DocumentOperations::Single(operation) => Ok(&operation.node),
                DocumentOperations::Multiple(map) if map.len() == 1 => {
                    let (_, operation) = map.iter().next().unwrap();
                    Ok(&operation.node)
                }
                DocumentOperations::Multiple(_) => Err(BuildError::OperationNameRequired),
            }
        }
    }

    #[inline(always)]
    pub fn build(
        &self,
        variables: &Variables<ConstValue>,
        operation_name: Option<&str>,
    ) -> Result<OperationPlan<ConstValue>, BuildError> {
        let mut fields = Vec::new();
        let mut fragments: HashMap<&str, &FragmentDefinition> = HashMap::new();

        for (name, fragment) in self.document.fragments.iter() {
            fragments.insert(name.as_str(), &fragment.node);
        }

        let operation = self.get_operation(operation_name)?;

        let name = self
            .get_type(operation.ty)
            .ok_or(BuildError::RootOperationTypeNotDefined { operation: operation.ty })?;
        fields.extend(self.iter(&operation.selection_set.node, name, None, &fragments));

        // skip the fields depending on variables.
        fields.retain(|f| !f.skip(variables));

        let plan = OperationPlan::new(fields, operation.ty);
        // TODO: operation from [ExecutableDocument] could contain definitions for
        // default values of arguments. That info should be passed to
        // [InputResolver] to resolve defaults properly
        let input_resolver = InputResolver::new(plan);

        Ok(input_resolver.resolve_input(variables)?)
    }
}

#[cfg(test)]
mod tests {
    use pretty_assertions::assert_eq;

    use super::*;
    use crate::core::blueprint::Blueprint;
    use crate::core::config::Config;
    use crate::core::jit::builder::Builder;
    use crate::core::valid::Validator;

    const CONFIG: &str = include_str!("./fixtures/jsonplaceholder-mutation.graphql");

    fn plan(
        query: impl AsRef<str>,
        variables: &Variables<ConstValue>,
    ) -> OperationPlan<ConstValue> {
        let config = Config::from_sdl(CONFIG).to_result().unwrap();
        let blueprint = Blueprint::try_from(&config.into()).unwrap();
        let document = async_graphql::parser::parse_query(query).unwrap();
        Builder::new(&blueprint, document)
            .build(variables, None)
            .unwrap()
    }

    #[tokio::test]
    async fn test_from_document() {
        let plan = plan(
            r#"
            query {
                posts { user { id name } }
            }
        "#,
            &Variables::new(),
        );
        assert!(plan.is_query());
        insta::assert_debug_snapshot!(plan.into_nested());
    }

    #[tokio::test]
    async fn test_size() {
        let plan = plan(
            r#"
            query {
                posts { user { id name } }
            }
        "#,
            &Variables::new(),
        );

        assert!(plan.is_query());
        assert_eq!(plan.size(), 4)
    }

    #[test]
    fn test_simple_query() {
        let plan = plan(
            r#"
            query {
                posts { user { id } }
            }
        "#,
            &Variables::new(),
        );

        assert!(plan.is_query());
        insta::assert_debug_snapshot!(plan.into_nested());
    }

    #[test]
    fn test_simple_mutation() {
        let plan = plan(
            r#"
            mutation {
              createUser(user: {
                id: 101,
                name: "Tailcall",
                email: "tailcall@tailcall.run",
                phone: "2345234234",
                username: "tailcall",
                website: "tailcall.run"
              }) {
                id
                name
                email
                phone
                website
                username
              }
            }
        "#,
            &Variables::new(),
        );

        assert!(!plan.is_query());
        insta::assert_debug_snapshot!(plan.into_nested());
    }

    #[test]
    fn test_fragments() {
        let plan = plan(
            r#"
            fragment UserPII on User {
              name
              email
              phone
            }

            query {
              user(id:1) {
                ...UserPII
              }
            }
        "#,
            &Variables::new(),
        );

        assert!(plan.is_query());
        insta::assert_debug_snapshot!(plan.into_nested());
    }

    #[test]
    fn test_multiple_operations() {
        let plan = plan(
            r#"
            query {
              user(id:1) {
                id
                username
              }
              posts {
                id
                title
              }
            }
        "#,
            &Variables::new(),
        );

        assert!(plan.is_query());
        insta::assert_debug_snapshot!(plan.into_nested());
    }

    #[test]
    fn test_variables() {
        let plan = plan(
            r#"
            query user($id: Int!) {
              user(id: $id) {
                id
                name
              }
            }
        "#,
            &Variables::from_iter([("id".into(), ConstValue::from(1))]),
        );

        assert!(plan.is_query());
        insta::assert_debug_snapshot!(plan.into_nested());
    }

    #[test]
    fn test_unions() {
        let plan = plan(
            r#"
            query {
              getUserIdOrEmail(id:1) {
                ...on UserId {
                  id
                }
                ...on UserEmail {
                  email
                }
              }
            }
        "#,
            &Variables::new(),
        );

        assert!(plan.is_query());
        insta::assert_debug_snapshot!(plan.into_nested());
    }

    #[test]
    fn test_default_value() {
        let plan = plan(
            r#"
            mutation {
              createPost(post:{
                userId:123,
                title:"tailcall",
                body:"tailcall test"
              }) {
                id
              }
            }
        "#,
            &Variables::new(),
        );

        assert!(!plan.is_query());
        insta::assert_debug_snapshot!(plan.into_nested());
    }

    #[test]
    fn test_condition() {
        // cases:
        // skip | include | ignore
        // T  |    T    |   T
        // T  |    F    |   T
        // T  |    V    |   T
        // F  |    F    |   T
        // F  |    T    |   F
        // F  |    V    |   F
        // V  |    T    |   F
        // V  |    F    |   F

        let test_var = Variable::new("ssdd.dev".to_string());

        let test_cases = vec![
            // ignore
            (Condition::True, Condition::True, true),
            (Condition::True, Condition::False, true),
            (Condition::True, Condition::Variable(test_var.clone()), true),
            (Condition::False, Condition::False, true),
            // don't ignore
            (Condition::False, Condition::True, false),
            (
                Condition::False,
                Condition::Variable(test_var.clone()),
                false,
            ),
            (
                Condition::Variable(test_var.clone()),
                Condition::True,
                false,
            ),
            (Condition::Variable(test_var), Condition::False, false),
        ];

        for (skip, include, expected) in test_cases {
            let conditions = Conditions { skip, include };
            assert_eq!(
                conditions.is_const_skip(),
                expected,
                "Failed for skip: {}, include: {}",
                conditions.skip,
                conditions.include
            );
        }
    }

    #[test]
    fn test_resolving_operation() {
        let query = r#"
            query GetPosts {
                posts {
                    id
                    userId
                    title
                }
            }

            mutation CreateNewPost {
                createPost(post: {
                    userId: 1,
                    title: "test-12",
                    body: "test-12",
                }) {
                    id
                    userId
                    title
                    body
                }
            }
        "#;
        let config = Config::from_sdl(CONFIG).to_result().unwrap();
        let blueprint = Blueprint::try_from(&config.into()).unwrap();
        let document = async_graphql::parser::parse_query(query).unwrap();
        let error = Builder::new(&blueprint, document.clone())
            .build(&Variables::new(), None)
            .unwrap_err();

        assert_eq!(error, BuildError::OperationNameRequired);

        let error = Builder::new(&blueprint, document.clone())
            .build(&Variables::new(), Some("unknown"))
            .unwrap_err();

        assert_eq!(error, BuildError::OperationNotFound("unknown".to_string()));

        let plan = Builder::new(&blueprint, document.clone())
            .build(&Variables::new(), Some("GetPosts"))
            .unwrap();
        assert!(plan.is_query());
        insta::assert_debug_snapshot!(plan.into_nested());

        let plan = Builder::new(&blueprint, document.clone())
            .build(&Variables::new(), Some("CreateNewPost"))
            .unwrap();
        assert!(!plan.is_query());
        insta::assert_debug_snapshot!(plan.into_nested());
    }

    #[test]
    fn test_directives() {
        let mut variables = Variables::new();
        variables.insert("includeName".to_string(), ConstValue::Boolean(true));

        let plan = plan(
            r#"
            query($includeName: Boolean! = true) {
                users {
                    id @options(paging: $includeName)
                    name @include(if: $includeName)
                }
            }
            "#,
            &variables,
        );

        assert!(plan.is_query());
        insta::assert_debug_snapshot!(plan.into_nested());
    }
}
