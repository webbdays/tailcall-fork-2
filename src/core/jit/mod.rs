mod exec;
mod model;
mod store;
mod synth;

use builder::*;
use store::*;
mod context;
mod error;
mod exec_const;
mod input_resolver;
mod request;
mod response;

// NOTE: Only used in tests and benchmarks
mod builder;
pub mod common;
mod graphql_executor;

// Public Exports
pub use error::*;
pub use exec_const::*;
pub use graphql_executor::*;
pub use model::*;
pub use request::*;
pub use response::*;
