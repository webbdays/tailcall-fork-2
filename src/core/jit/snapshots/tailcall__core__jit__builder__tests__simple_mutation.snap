---
source: src/core/jit/builder.rs
expression: plan.into_nested()
---
[
    Field {
        id: 0,
        name: "createUser",
        ir: "Some(..)",
        type_of: User,
        args: [
            Arg {
                id: 0,
                name: "user",
                type_of: InputUser!,
                value: Some(
                    Object(
                        {
                            Name(
                                "id",
                            ): Number(
                                Number(101),
                            ),
                            Name(
                                "name",
                            ): String(
                                "Tailcall",
                            ),
                            Name(
                                "email",
                            ): String(
                                "tailcall@tailcall.run",
                            ),
                            Name(
                                "phone",
                            ): String(
                                "2345234234",
                            ),
                            Name(
                                "username",
                            ): String(
                                "tailcall",
                            ),
                            Name(
                                "website",
                            ): String(
                                "tailcall.run",
                            ),
                        },
                    ),
                ),
                default_value: None,
            },
        ],
        extensions: Some(
            Nested(
                [
                    Field {
                        id: 1,
                        name: "id",
                        type_of: ID!,
                        is_scalar: true,
                        directives: [],
                    },
                    Field {
                        id: 2,
                        name: "name",
                        type_of: String!,
                        is_scalar: true,
                        directives: [],
                    },
                    Field {
                        id: 3,
                        name: "email",
                        type_of: String!,
                        is_scalar: true,
                        directives: [],
                    },
                    Field {
                        id: 4,
                        name: "phone",
                        type_of: String,
                        is_scalar: true,
                        directives: [],
                    },
                    Field {
                        id: 5,
                        name: "website",
                        type_of: String,
                        is_scalar: true,
                        directives: [],
                    },
                    Field {
                        id: 6,
                        name: "username",
                        type_of: String!,
                        is_scalar: true,
                        directives: [],
                    },
                ],
            ),
        ),
        is_scalar: false,
        directives: [],
    },
]
